sound/components/virt.c

```diff
@@ -0,0 +1,263 @@
+
+static int owners[MAX_STREAMS];
+
+static void respond_to_cmd(sound_queues_t *client_queues,
+                           sound_cmd_t *cmd,
+                           sound_status_t status)
+{
+    cmd->status = status;
+    if (sound_enqueue_cmd(&client_queues->cmd_res, cmd) != 0) {
+        sddf_dprintf("SND VIRT|ERR: failed to respond to command\n");
+    }
+}
+
+static void respond_to_pcm(sound_queues_t *client_queues,
+                           sound_pcm_t *pcm,
+                           sound_status_t status)
+{
+    pcm->status = status;
+    pcm->latency_bytes = 0;
+    if (sound_enqueue_pcm(&client_queues->pcm_res, pcm) != 0) {
+        sddf_dprintf("SND VIRT|ERR: failed to respond to pcm\n");
+    }
+}
+
+static int notified_by_client(int client)
+{
+    if (client < 0 || client > NUM_CLIENTS) {
+        sddf_dprintf("SND VIRT|ERR: invalid client id %d\n", client);
+        return -1;
+    }
+
+    bool notify_client = false;
+    bool notify_driver = false;
+    sound_queues_t *client_queues = &clients[client];
+    sound_cmd_t cmd;
+
+    while (sound_dequeue_cmd(&client_queues->cmd_req, &cmd) == 0) {
+
+        if (cmd.stream_id > MAX_STREAMS) {
+            sddf_dprintf(
+                "SND VIRT|ERR: [client %d] stream id %u too large (max %u)\n",
+                client, cmd.stream_id, MAX_STREAMS);
+            respond_to_cmd(client_queues, &cmd, SOUND_S_BAD_MSG);
+            continue;
+        }
+
+        if (owners[cmd.stream_id] == NO_OWNER) {
+            if (cmd.code == SOUND_CMD_TAKE) {
+                owners[cmd.stream_id] = client;
+            } else {
+                sddf_dprintf("SND VIRT|ERR: [client %d] client must take first\n",
+                             client);
+                respond_to_cmd(client_queues, &cmd, SOUND_S_BAD_MSG);
+                notify_client = true;
+                continue;
+            }
+        }
+
+        int owner = owners[cmd.stream_id];
+        if (owner != client) {
+            sddf_dprintf("SND VIRT|ERR: [client %d] stream busy\n", client);
+            respond_to_cmd(client_queues, &cmd, SOUND_S_BUSY);
+            notify_client = true;
+            continue;
+        }
+
+        if (sound_enqueue_cmd(&driver_queues.cmd_req, &cmd) != 0) {
+            sddf_dprintf(
+                "SND VIRT|ERR: [client %d] failed to enqueue command\n",
+                client);
+            return -1;
+        }
+        notify_driver = true;
+    }
+
+    sound_pcm_t pcm;
+    while (sound_dequeue_pcm(&client_queues->pcm_req, &pcm) == 0) {
+
+        if (pcm.stream_id > MAX_STREAMS) {
+            sddf_dprintf("SND VIRT|ERR: [client %d] stream id too large\n", client);
+            respond_to_pcm(client_queues, &pcm, SOUND_S_BAD_MSG);
+            notify_client = true;
+            continue;
+        }
+
+        int owner = owners[pcm.stream_id];
+        if (owner != client) {
+            sddf_dprintf("SND VIRT|ERR: [client %d] driver replied to busy stream\n",
```

Comment:
```suggestion
            sddf_dprintf("SND VIRT|ERR: client %d driver requested busy stream\n",
```

also print out stream ID